overlord/snapstate/handlers.go

```diff
@@ -821,17 +821,38 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
-// maybeRestart will schedule a reboot or restart as needed for the just linked
-// snap with info if it's a core or kernel snap.
+// maybeRestart will schedule a reboot or restart as needed for the
+// just linked snap with info if it's a core or snapd or kernel snap.
 func maybeRestart(t *state.Task, info *snap.Info) {
 	st := t.State()
+
+	// TODO: once classic uses the snapd snap we need to restart
+	//       here too
 	if release.OnClassic && info.Type == snap.TypeOS {
 		t.Logf("Requested daemon restart.")
 		st.RequestRestart(state.RestartDaemon)
 	}
-	if !release.OnClassic && boot.KernelOsBaseRebootRequired(info) {
-		t.Logf("Requested system restart.")
-		st.RequestRestart(state.RestartSystem)
+
+	if !release.OnClassic {
+		// On a core system we may need a full reboot if
+		// core/base or the kernel changes.
+		if boot.KernelOsBaseRebootRequired(info) {
```

Comment:
Would be nice to improve this name to be more clear.. the implied "if" is getting out of hand. :-)

Perhaps something like ChangeRequiresReboot?

Side note: that function has a typo in its documentation ("swith").